library-best-practices/glossary.md

```diff
@@ -0,0 +1,84 @@
+Java Dependency Glossary
+------------------------
+
+### Types of conflicts and compatibility
+
+- **Linkage conflict**: The signature of a non-private method or class in a
+  dependency has changed in an incompatible way between the version supplied at
+  compile time and the version invoked at runtime. For example, a public method
+  may be removed from a class or a class may be made final. Linkage conflicts
+  detected at runtime manifest as ReflectiveOperationException,
+  `NoClassDefFoundError`, `NoSuchFieldException`, `MethodNotFoundException`,
+  `LinkageError`, or other related exceptions.
+  - Or, another perspective: In cases where binary compatibility and source
+    compatibility are the same, a linkage conflict is when compilation would
+    fail if the libraries in the classpath were all built together from their
+    originating source code, or when reflection would fail.
+  - Opposite: **Linkage-compatible**.
+  - Sub-type: **Static linkage conflict**: A linkage conflict caused by a direct
+    code reference (non-reflective).
+  - Sub-type: **Reflective linkage conflict**: A linkage conflict caused by a
+    reflective reference.
+- **Behavior conflict**: The class's implementation has changed in a way that
+  can break clients at runtime although all signatures remain compatible. For
+  example, if a method that has been returning mutable lists begins returning
+  immutable lists without updating its signature, dependents that mutate the
+  list will fail, possibly far away from the original call site. By contrast, a
+  change in return type from List to ImmutableList would be a linkage conflict.
+  - Opposite: **Behavior-compatible**.
+- **Version disagreement**: This is a class of failure reported by build systems
+  with strict version checking on. The main example is dependency convergence
+  failures when using Maven version enforcer, which checks that all branches of
+  a dependency tree agree on the version of every dependency. A failure of
+  dependency convergence doesn't guarantee a linkage or behavior conflict, and
+  the passing of a dependency convergence check doesn't guarantee there will be
```

Comment:
If dependencies converge, static linkage errors will be detected at compile time. 

Reflective errors and behavior conflicts at this point indicate inadequate test coverage.

Reply:
The passing of a dependency convergence check != dependencies converging. As noted elsewhere, dependency convergence checks will succeed when dependencyManagement is used, even if the tree still references multiple versions. Also, if exclusions are used so that the tree only references a single version per dependency and the dependency convergence check passes, the library trying to reference the excluded dependency can try to pull in a more recent version and use new features, and then there will be static linkage errors because the dependency tree is using an older version. In the theoretical scenario that the entire tree is compiled with identical dependencies and those are the same as the versions used in the final dependency tree, *only then* will you be guaranteed that you will have no static linkage errors at runtime. However, the text doesn't reference that idea, it only references the attempted check for convergence.

All this to say - the statement in the text still stands as-is.